### ISSUE TITLE
drill+jdbc: ModuleNotFoundError: No module named 'jpype.dbapi2'

### ISSUE THREAD

**Issue by gdubya** [JohnOmernik/sqlalchemy-drill#53]
I've successfully connected to Drill using the "standard" connection:

```
drill = create_engine("drill+sadrill://localhost:8047/dfs?use_ssl=False")
```

But this has a problem when selecting a large dataset:

```
sqlalchemy_drill.drilldbapi.api_exceptions.ProgrammingError: 'RESOURCE ERROR: There is not enough heap memory to run this query using the web interface. \n\nPlease try a query with fewer columns or with a filter or limit condition to limit the data returned. \nYou can also try an ODBC/JDBC client. \n\n[Error Id: 8961c69f-e237-4c92-8d69-0e0d61aeabac ] HTTP ERROR: 500'
```

So i'm trying to use the JDBC driver:

```
drill = create_engine("drill+jdbc://localhost")
```

But then I get a ModuleNotFoundError.

```
ModuleNotFoundError: No module named 'jpype.dbapi2'
```

I have jpype 1.0.2 installed. What could be the cause?

**Comment by gdubya** [comment]
Cross posted t[o jpype](https://github.com/jpype-project/jpype/issues/859). Looks like Jpype >= 1.0.3 is required.

**Comment by cgivre** [comment]
@gdubya Does it work with `Jpype` >= 1.03?

**Comment by gdubya** [comment]
It's not released yet, but the message in that other issue was that dbapi is part of jpype 1.1 (which is possibly the next release after the current, which is 1.0.2). I don't know if they'll do a 1.0.3 or 1.1. I guess i should have said "Jpype > 1.0.2".

But... am I wrong about that? the README says Jpype 0.6.3. Maybe the API was removed after that version and is going to be reintroduced in the next version.

EDIT: Hmm no, same error with 0.6.3.

**Comment by Thrameos** [comment]
DBAPI was introduced in 1.0.3 dev which because of changes to the external exposed API (such as introducing dbapi2) is to be released as 1.1.0.   The release is in process and should be on the distribution sites shortly baring unexpected problems.  We have a very large test bench but every so often problematic code does slip through forcing patch releases. 

No version prior to JPype 1.0.x should be used in production code.   Many of the earlier JPype versions have severe memory leaks or other undesirable behaviors.   One of the major reasons for creating the dbapi2 interface in JPype is that JayDeBeAPI users keep recommending rolling back to earlier JPype versions to avoid maintenance issues within JayDeBeAPI.   This in no way means that I am discouraging use of JayDeBeAPI when a project needs to work with Jython and CPython equally well as there are clear use cases where that would be desirable.  But those project should keep up to date with the latest JPype, and if JayDeBeAPI does not work with the latest that should be addressed with issue tickets on JayDeBeAPI.

DBAPI2 support in JPype is somewhat a work in progress.  I am not a heavy database user and therefore I merely guessed at features that a dbapi2 user would likely need.   Things like exposing exotic data types and manual adapters and converters are new.  I executed all of the dbapi2 conformance tests that I could locate (resolving some differences between the test benches).  I also filled out many of the stubs that JayDeBeAPI had skipped so I suspect that it will work better with frameworks such as drill, but testing is badly needed.  I will issue a patch fix for issue if they are reported in the next few weeks to try make sure it is usable to the community of jdbc users.
